src/logic/hydra.lean

```diff
@@ -121,10 +119,22 @@ variable (r : α → α → Prop)
 
   The lemma `relation.cut_expand_iff` below converts between this convenient definition
   and the direct translation when `r` is irreflexive. -/
-def cut_expand (s' s : multiset α) : Prop :=
+def cut_expand (r : α → α → Prop) (s' s : multiset α) : Prop :=
 ∃ (t : multiset α) (a : α), (∀ a' ∈ t, r a' a) ∧ s' + {a} = s + t
 
-lemma cut_expand_iff [decidable_eq α] {r} (hr : irreflexive r) {s' s : multiset α} :
+variable {r : α → α → Prop}
+
+theorem cut_expand_singleton {s x} (h : ∀ x' ∈ s, r x' x) : cut_expand r s {x} :=
+⟨s, x, h, add_comm s _⟩
+
+theorem cut_expand_singleton_singleton {x' x} (h : r x' x) : cut_expand r {x'} {x} :=
+cut_expand_singleton (λ a h, by rwa mem_singleton.1 h)
+
+theorem cut_expand_add_of_cut_expand {t u} (s) :
+  cut_expand r t u → cut_expand r (s + t) (s + u) :=
+λ ⟨v, a, h, h'⟩, ⟨v, a, h, by rw [add_assoc, h', ←add_assoc]⟩
```

Comment:
This could be made an `iff` without much effort:
```suggestion
theorem cut_expand_add_left {t u} (s) : cut_expand r (s + t) (s + u) ↔ cut_expand r t u :=
⟨λ ⟨v, a, h, h'⟩, ⟨v, a, h, by { rw [add_assoc, add_assoc] at h', exact add_left_cancel h' }⟩,
 λ ⟨v, a, h, h'⟩, ⟨v, a, h, by rw [add_assoc, h', ←add_assoc]⟩⟩
```